sw/cheri/boot/boot_loader.cc

```diff
@@ -14,86 +16,81 @@
 
 const char prefix[] = "\x1b[35mbootloader\033[0m: ";
 
-struct UF2Block
-{
-	uint64_t magicStart;
-	uint32_t flags;
-	uint32_t targetAddr;
-	uint32_t payloadSize;
-	uint32_t blockNo;
-	uint32_t numBlocks;
-	uint32_t fileSizeOrFamilyId;
-	uint32_t data[119];
-	uint32_t magicEnd;
-
-	static constexpr size_t size = 512;
-
-	void read_from_flash(SpiFlash &flash, uint32_t address)
-	{
-		flash.read(address, (uint8_t *)this, size);
-	}
-
-	auto check_magic() -> bool
-	{
-		return 0x9e5d'5157'0a32'4655 == magicStart && 0x0ab1'6f30 == magicEnd;
-	}
-};
-
-static_assert(sizeof(UF2Block) == UF2Block::size,
-              "The UF2Block structure is the wrong size.");
-
 typedef CHERI::Capability<volatile OpenTitanUart<>> &UartRef;
 
 [[noreturn]] void complain_and_loop(UartRef uart, const char *str)
 {
 	write_str(uart, prefix);
 	write_str(uart, str);
-	while (true) {
-		asm ("wfi");
+	while (true)
+	{
+		asm("wfi");
 	}
 }
 
-void read_blocks(SpiFlash                   &flash,
-                 UartRef                     uart,
-                 CHERI::Capability<uint32_t> sram)
+void read_elf(SpiFlash &flash, UartRef uart, CHERI::Capability<uint32_t> sram)
 {
 	write_str(uart, prefix);
 	write_str(uart, "Loading software from flash...\r\n");
 
-	UF2Block block;
-	block.read_from_flash(flash, 0x0);
+	Elf32_Ehdr ehdr;
+	flash.read(0x0, (uint8_t *)&ehdr, sizeof(Elf32_Ehdr));
+
+	// Check the ELF magic numbers.
+	if (ehdr.e_ident[EI_MAG0] != ELFMAG0 || ehdr.e_ident[EI_MAG1] != ELFMAG1 ||
+	    ehdr.e_ident[EI_MAG2] != ELFMAG2 || ehdr.e_ident[EI_MAG3] != ELFMAG3)
+	{
+		complain_and_loop(uart, "Failed ELF Magic Check\r\n");
+	}
 
-	if (!block.check_magic())
+	if (ehdr.e_ident[EI_CLASS] != ELFCLASS32 || ehdr.e_type != ET_EXEC ||
+	    ehdr.e_machine != EM_RISCV ||
+	    (ehdr.e_flags & (EF_RISCV_CHERIABI | EF_RISCV_CAP_MODE)) !=
+	      (EF_RISCV_CHERIABI | EF_RISCV_CAP_MODE))
 	{
-		complain_and_loop(uart, "Failed Magic Check\r\n");
+		complain_and_loop(uart,
+		                  "ELF file is not 32-bit CHERI RISC-V executable\r\n");
 	}
-	auto write_block = [sram](UF2Block &block) {
 
-		auto block_addrs      = sram;
-		block_addrs.address() = block.targetAddr;
-		block_addrs.bounds()  = block.payloadSize;
-		const size_t words    = block.payloadSize / 4;
+	write_str(uart, prefix);
+	write_str(uart, "Offset   VirtAddr FileSize MemSize\r\n");
 
-		for (uint32_t word = 0; word < words; ++word)
+	Elf32_Phdr phdr;
+	for (uint32_t i = 0; i < ehdr.e_phnum; i++)
+	{
+		flash.read(ehdr.e_phoff + ehdr.e_phentsize * i,
+		           (uint8_t *)&phdr,
+		           sizeof(Elf32_Phdr));
+
+		if (phdr.p_type != PT_LOAD)
+			continue;
+
+		write_str(uart, prefix);
+		write_hex(uart, phdr.p_offset);
+		write_str(uart, " ");
+		write_hex(uart, phdr.p_vaddr);
+		write_str(uart, " ");
+		write_hex(uart, phdr.p_filesz);
+		write_str(uart, " ");
+		write_hex(uart, phdr.p_memsz);
+		write_str(uart, "\r\n");
```

Comment:
Please can we remove this table from the output logs?

Reply:
What's the issue of having the table?

Reply:
It's unnecessary noise in the output

Reply:
Gated under a macro